# Practical 2: Numerical Integration of Chemical Kinetics using Python Loops

### Computer Programming Lab – Chemical Engineering
**Topic:** Looping Constructs (`for`, `while`) with Numerical Integration of Kinetics  
**Environment:** GitHub Codespaces + Jupyter Notebook + GitHub Copilot

---

## Important: How we will use AI Copilot

In this practical, GitHub Copilot is your **AI coding assistant**. Use it to understand loops, numerical integration, debugging and testing.

**Do not ask Copilot to write the complete practical at the beginning.** First understand the chemical equation, convert it into a numerical calculation, and decide what must be repeated. Then use Copilot for hints, explanations and debugging.

At the end, you should be able to explain your own program without depending on Copilot.


## 1. Learning Objectives

After completing this practical, you should be able to:

1. Understand why loops are useful for repeated chemical-engineering calculations.
2. Understand the basic idea of numerical integration using small time steps.
3. Apply the Euler method to a simple first-order reaction.
4. Use a `for` loop to perform a fixed number of integration steps.
5. Use a `while` loop when the number of steps depends on a process condition.
6. Calculate concentration and conversion during a reaction.
7. Investigate how the time step affects the numerical result.
8. Use GitHub Copilot to explain, debug and improve a numerical simulation.


## 2. Chemical Engineering Context

Chemical engineers frequently deal with equations that describe how a quantity changes with time. For example, the concentration of a reactant in a batch reactor changes as a reaction proceeds.

Consider the simple first-order reaction:

```text
A → Products
```

For a first-order reaction, the rate of disappearance of A can be represented by:

```text
dC/dt = -kC
```

where:

- `C` = concentration of reactant A (mol/L)
- `t` = time (min)
- `k` = first-order rate constant (1/min)

Instead of solving the differential equation analytically, we will use **numerical integration**. This means that we calculate the concentration step-by-step over small time intervals.

**Note:** This is an educational numerical model for programming practice. It is not a substitute for a rigorous reactor-design calculation.


## 3. Euler Method – The Basic Idea

For a small time step `Δt`, the Euler method approximates the next concentration using:

```text
C_new = C_old + (dC/dt) × Δt
```

Since:

```text
dC/dt = -kC
```

we obtain:

```text
C_new = C_old - k × C_old × Δt
```

The important programming idea is that this calculation is repeated many times.


## 4. Example Problem

A batch reactor initially contains reactant A at:

- Initial concentration: `C₀ = 1.00 mol/L`
- Rate constant: `k = 0.05 1/min`
- Time step: `Δt = 1 min`
- Total simulation time: `10 min`

Use the Euler method to calculate the concentration at every minute.

Your program should produce a sequence similar to:

```text
Time       Concentration
0          1.0000
1          ...
2          ...
...
10         ...
```


## 5. Activity 1 – Understand What Must Be Repeated

Before writing Python code, identify the repeated calculation.

At every time step, you need to:

1. Calculate the rate `dC/dt`.
2. Calculate the change in concentration.
3. Calculate the new concentration.
4. Increase the time.

### Your task

Write in your own words: **Why is a loop suitable for this problem?**


### Copilot Activity 1 – Understand the Loop

Ask Copilot:

> I am a second-year chemical engineering student learning Python. I need to repeat an Euler-method calculation for a first-order reaction at several time steps. Explain why a for loop is suitable. Do not give me the complete program.

Write one or two important points you learned from Copilot.


In [1]:
# Activity 1
# Print the time values from 0 to 10 minutes using a for loop.
# Write your solution here.
t= 15
for i in range(t+1):
    print(f"{i} minutes")       
print("Activity 1 completed")        

0 minutes
1 minutes
2 minutes
3 minutes
4 minutes
5 minutes
6 minutes
7 minutes
8 minutes
9 minutes
10 minutes
11 minutes
12 minutes
13 minutes
14 minutes
15 minutes
Activity 1 completed


## 6. Activity 2 – Implement One Euler Integration Step

Before creating a loop, implement **one** numerical integration step.

Use:

- `C = 1.00 mol/L`
- `k = 0.05 1/min`
- `dt = 1 min`

Calculate:

1. Reaction rate `dC/dt`
2. New concentration `C_new`

This helps you understand the calculation before repeating it.


### Copilot Activity 2 – Ask for Explanation, Not the Answer

Ask Copilot:

> Explain how to translate dC/dt = -kC into one Euler integration step in Python. I want to understand the variables and sequence of calculations. Do not give me a complete simulation program.


In [2]:
# Activity 2
# Calculate one Euler step.
# Start with C = 1.00, k = 0.07 and dt = 1.
c = 1.00
k = 0.07
dt = 1
c_new = c - k * c * dt
print(f"New concentration after one Euler step: {c_new:.4f}")
print("Activity 2 completed")

New concentration after one Euler step: 0.9300
Activity 2 completed


## 7. Activity 3 – Numerical Integration using a `for` Loop

Now place the Euler calculation inside a `for` loop.

Perform 10 integration steps.

At each iteration:

1. Calculate `dC/dt`.
2. Update `C` using the Euler method.
3. Increase time by `dt`.
4. Display time and concentration.

### Important

The concentration calculated in one step becomes the starting concentration for the next step.


### Copilot Activity 3 – Debugging/Hint Prompt

If you are stuck, ask:

> I am implementing Euler integration for dC/dt = -kC inside a Python for loop. Explain what should be updated after each iteration and why. Do not write the complete code.


In [3]:
# Activity 3
# Use a for loop to numerically integrate the first-order reaction for 10 minutes.
c = 1.00
k = 0.05
dt = 1
t= 10
for i in range(t+1):
    #print(f": {i} minutes") 
    print(f"Concentration at {i} minutes: {c:.4f}")
    c = c - k * c * dt
print("Activity 3 completed")

Concentration at 0 minutes: 1.0000
Concentration at 1 minutes: 0.9500
Concentration at 2 minutes: 0.9025
Concentration at 3 minutes: 0.8574
Concentration at 4 minutes: 0.8145
Concentration at 5 minutes: 0.7738
Concentration at 6 minutes: 0.7351
Concentration at 7 minutes: 0.6983
Concentration at 8 minutes: 0.6634
Concentration at 9 minutes: 0.6302
Concentration at 10 minutes: 0.5987
Activity 3 completed


## 8. Activity 4 – Calculate Conversion

For a batch reaction, conversion can be calculated as:

```text
Conversion (%) = ((C₀ - C) / C₀) × 100
```

Modify your simulation so that each time step displays:

- Time
- Concentration
- Conversion

Think about whether conversion should be calculated using the **initial concentration** or the current concentration.


### Copilot Activity 4 – Explain Conversion

Ask Copilot:

> Explain how conversion is calculated from initial and current reactant concentration in a batch reactor. Explain what happens to conversion as the reactant concentration decreases. Do not write the Python program.


In [4]:
# Activity 4
# Add conversion (%) to your numerical integration program.
c = 1.00
k = 0.05
dt = 1
t= 10
conversion = 0.0
for i in range(t+1):
    conversion = (1 - c) * 100
    print(f"Concentration at {i} minutes: {c:.4f}, Conversion: {conversion:.4f}%")
    #print(f": {i} minutes") 
    dcdt = -k * c     
    c = c - k * c * dt
    conversion = (1 - c) * 100
print(f"Concentration after 10  minutes: {conversion:.4f}%")

Concentration at 0 minutes: 1.0000, Conversion: 0.0000%
Concentration at 1 minutes: 0.9500, Conversion: 5.0000%
Concentration at 2 minutes: 0.9025, Conversion: 9.7500%
Concentration at 3 minutes: 0.8574, Conversion: 14.2625%
Concentration at 4 minutes: 0.8145, Conversion: 18.5494%
Concentration at 5 minutes: 0.7738, Conversion: 22.6219%
Concentration at 6 minutes: 0.7351, Conversion: 26.4908%
Concentration at 7 minutes: 0.6983, Conversion: 30.1663%
Concentration at 8 minutes: 0.6634, Conversion: 33.6580%
Concentration at 9 minutes: 0.6302, Conversion: 36.9751%
Concentration at 10 minutes: 0.5987, Conversion: 40.1263%
Concentration after 10  minutes: 43.1200%


## 9. Activity 5 – Numerical Integration with a `while` Loop

A `for` loop is convenient when the number of integration steps is known.

Now consider a different engineering requirement:

> Continue the reaction simulation until conversion reaches **80%**, but stop if the simulation time reaches **100 minutes**.

The number of steps is not known in advance, so a `while` loop is appropriate.

Your loop should:

1. Calculate the reaction rate.
2. Perform one Euler step.
3. Update time.
4. Calculate conversion.
5. Continue while the target has not been reached and maximum time has not been exceeded.


### Copilot Activity 5 – Understand `while`

Ask Copilot:

> Explain how to design a while-loop condition for a chemical reaction simulation that should continue until conversion reaches 80% or time reaches 100 minutes. Explain the logic without writing the complete program.


In [5]:
# Activity 5
# Use a while loop for the target-conversion simulation.
t= 0
while t<= 10:
    print(f"{t} minutes")
    t = t+1
print("Activity 5 completed")    

0 minutes
1 minutes
2 minutes
3 minutes
4 minutes
5 minutes
6 minutes
7 minutes
8 minutes
9 minutes
10 minutes
Activity 5 completed


In [6]:
c = 1.00
k = 0.05
dt = 1
t=0 
conversion = 0.0
while (conversion <=80.0 and t <= 100):
    print(f"Concentration at {t} minutes: {c:.4f}, Conversion: {conversion:.2f}%")
    dcdt = -k * c
    c = c - k * c * dt
    conversion = (1 - c) * 100
    t = t + 1
    

Concentration at 0 minutes: 1.0000, Conversion: 0.00%
Concentration at 1 minutes: 0.9500, Conversion: 5.00%
Concentration at 2 minutes: 0.9025, Conversion: 9.75%
Concentration at 3 minutes: 0.8574, Conversion: 14.26%
Concentration at 4 minutes: 0.8145, Conversion: 18.55%
Concentration at 5 minutes: 0.7738, Conversion: 22.62%
Concentration at 6 minutes: 0.7351, Conversion: 26.49%
Concentration at 7 minutes: 0.6983, Conversion: 30.17%
Concentration at 8 minutes: 0.6634, Conversion: 33.66%
Concentration at 9 minutes: 0.6302, Conversion: 36.98%
Concentration at 10 minutes: 0.5987, Conversion: 40.13%
Concentration at 11 minutes: 0.5688, Conversion: 43.12%
Concentration at 12 minutes: 0.5404, Conversion: 45.96%
Concentration at 13 minutes: 0.5133, Conversion: 48.67%
Concentration at 14 minutes: 0.4877, Conversion: 51.23%
Concentration at 15 minutes: 0.4633, Conversion: 53.67%
Concentration at 16 minutes: 0.4401, Conversion: 55.99%
Concentration at 17 minutes: 0.4181, Conversion: 58.19%
Conce

## 10. Activity 6 – Investigate the Effect of Time Step

The Euler method is a numerical approximation. The size of `dt` can affect the result.

Run the same reaction using:

- `dt = 2 min`
- `dt = 1 min`
- `dt = 0.5 min`
- `dt = 0.1 min`

Compare the concentration obtained at approximately 10 minutes.

### Think about it

Does reducing the time step make the result change? Why might a smaller time step generally give a better numerical approximation?


### Copilot Activity 6 – Understand Numerical Error

Ask Copilot:

> Explain in beginner-friendly terms why the time-step size can affect an Euler numerical integration result. Use a first-order chemical reaction as the example. Do not provide code.


In [7]:
# Activity 6
# Compare Euler integration results for different dt values.
c= 1.00
k= 0.05
dt_values = [1, 0.5, 0.1]
for dt in dt_values:
    c = 1.00
    t= 0
    print(f"\nResults for dt = {dt}:")
    while t <= 10:
         dCdt = -k * c
         c = c - k * c * dt 
         conversion = (1 - c) * 100
         print(f"Time {t:.2f} minutes, Concentration: {c:.4f}, Conversion: {conversion:.2f}%")
         t += dt


Results for dt = 1:
Time 0.00 minutes, Concentration: 0.9500, Conversion: 5.00%
Time 1.00 minutes, Concentration: 0.9025, Conversion: 9.75%
Time 2.00 minutes, Concentration: 0.8574, Conversion: 14.26%
Time 3.00 minutes, Concentration: 0.8145, Conversion: 18.55%
Time 4.00 minutes, Concentration: 0.7738, Conversion: 22.62%
Time 5.00 minutes, Concentration: 0.7351, Conversion: 26.49%
Time 6.00 minutes, Concentration: 0.6983, Conversion: 30.17%
Time 7.00 minutes, Concentration: 0.6634, Conversion: 33.66%
Time 8.00 minutes, Concentration: 0.6302, Conversion: 36.98%
Time 9.00 minutes, Concentration: 0.5987, Conversion: 40.13%
Time 10.00 minutes, Concentration: 0.5688, Conversion: 43.12%

Results for dt = 0.5:
Time 0.00 minutes, Concentration: 0.9750, Conversion: 2.50%
Time 0.50 minutes, Concentration: 0.9506, Conversion: 4.94%
Time 1.00 minutes, Concentration: 0.9269, Conversion: 7.31%
Time 1.50 minutes, Concentration: 0.9037, Conversion: 9.63%
Time 2.00 minutes, Concentration: 0.8811, Conv

## 11. Activity 7 – Compare with the Analytical Solution

For a first-order reaction, the analytical solution is:

```text
C(t) = C₀ exp(-kt)
```

Use Python's `math.exp()` to calculate the analytical concentration at selected times.

Then compare it with your Euler result.

Calculate the absolute error:

```text
Absolute Error = |Euler concentration - Analytical concentration|
```

This activity helps you understand that numerical integration is an **approximation**.


### Copilot Activity 7 – Comparison and Explanation

Ask Copilot:

> Explain the difference between a numerical Euler solution and the analytical solution of a first-order reaction. Why can their values be different? Do not write the complete comparison program.


In [8]:
# Activity 7
# Compare Euler and analytical concentrations at selected times.


## 12. Copilot Debugging Challenge

The following program contains a common numerical-integration error.

Study it before asking Copilot.

```python
C = 1.0
k = 0.05
dt = 1.0

for step in range(10):
    rate = -k * C
    C_new = C + rate * dt
    print(step + 1, C)
```

### Your task

1. Run the program.
2. Look carefully at the concentration values.
3. Decide whether the concentration is actually being updated.
4. Ask Copilot:

> My Euler integration program calculates C_new but the next iteration does not appear to use the new concentration. Explain the likely logic problem without rewriting the complete program.

5. Correct the program yourself.


In [9]:
# Run the intentionally incorrect program here.
C = 1.0
k = 0.05
dt = 1.0

for step in range(10):
    rate = -k * C
    C_new = C + rate * dt
    print(step + 1, C)


1 1.0
2 1.0
3 1.0
4 1.0
5 1.0
6 1.0
7 1.0
8 1.0
9 1.0
10 1.0


### Explain the error in your own words

Write your explanation here before continuing.


## 13. Testing Your Numerical Integration Program

Test your program using different reaction conditions.

| Test | C₀ (mol/L) | k (1/min) | dt (min) | Total Time (min) |
|---|---:|---:|---:|---:|
| 1 | 1.00 | 0.05 | 1.0 | 10 |
| 2 | 2.00 | 0.02 | 1.0 | 20 |
| 3 | 1.50 | 0.10 | 0.5 | 10 |
| 4 | 1.00 | 0.05 | 0.1 | 10 |
| 5 | 1.00 | 0.05 | 2.0 | 10 |

For each case, record the final Euler concentration and, where possible, compare it with the analytical solution.


## 14. Final Mini-Project – Batch Reaction Numerical Integrator

Combine your work into one clean Python program.

The final program should:

1. Accept `C₀`, `k`, `dt` and total simulation time from the user.
2. Validate the input values.
3. Numerically integrate `dC/dt = -kC` using the Euler method.
4. Use a loop to perform the integration steps.
5. Display time, concentration and conversion.
6. Calculate the analytical concentration using `math.exp()`.
7. Display the numerical error at the final time.
8. Clearly state that the Euler result is a numerical approximation.

**Keep the program beginner-friendly. Do not use advanced libraries or techniques that have not been discussed in class.**


In [10]:
# FINAL PROGRAM
# Batch Reaction Numerical Integrator
# Write your clean final solution here.


## 15. Final Copilot Review

After completing your final program, ask Copilot:

> Review my Python Euler integration program for a first-order chemical reaction. Check whether the loop updates concentration correctly, whether the time step is used correctly, whether the stopping condition is correct, and whether the numerical result is compared properly with the analytical solution. Do not rewrite my program; identify possible issues and explain them.

Make changes only after you understand the feedback.


## 16. Reflection – What Did You Learn?

Answer in your own words:

1. Why is a loop required for numerical integration?
2. What does `dC/dt = -kC` represent in this practical?
3. What does one Euler step calculate?
4. Why does the concentration from one step become the starting concentration for the next step?
5. What is the role of `dt` in numerical integration?
6. What did you observe when `dt` was changed?
7. Why can the Euler result differ from the analytical result?
8. What error did Copilot help you understand?
9. Which part of your program did you write or correct yourself after receiving a Copilot hint?
10. Can you explain your final program without opening Copilot?


## 17. Submission Checklist

- [ ] I completed all activities.
- [ ] I used a `for` loop for numerical integration.
- [ ] I used a `while` loop for condition-controlled integration.
- [ ] I implemented the Euler method correctly.
- [ ] I calculated conversion.
- [ ] I investigated different time-step sizes.
- [ ] I compared numerical and analytical results.
- [ ] I calculated/observed numerical error.
- [ ] I tested different reaction conditions.
- [ ] I used Copilot for explanation and debugging rather than simply copying a complete solution.
- [ ] I can explain my final program.
- [ ] I saved the completed notebook in my GitHub repository.


## 18. Suggested Evaluation – Faculty Use

| Component | Marks |
|---|---:|
| Basic variables and input | 1 |
| Correct implementation of Euler step | 2 |
| Correct use of loop for integration | 2 |
| Conversion calculation | 1 |
| Time-step investigation / numerical error | 1 |
| Numerical vs analytical comparison | 1 |
| Copilot usage and explanation | 1 |
| Student's ability to explain/modify the program | 1 |
| **Total** | **10** |

**Faculty note:** During evaluation, change `dt` or `k` and ask the student to predict what should happen. Also ask why the concentration from the previous step is required in the next Euler calculation.
